daemon/daemon.go

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2022, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package daemon
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/megaease/easeprobe/global"
+)
+
+// Config is the daemon config
+type Config struct {
+	PIDFile string
+	pidFd   *os.File
+}
+
+// NewPIDFile create a new pid file
+func NewPIDFile(pidfile string) (*Config, error) {
+	c := &Config{
+		PIDFile: pidfile,
+	}
+	_, err := c.CheckPIDFile()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(c.PIDFile), 0755); err != nil {
```

Comment:
Do we really need to create the entire path tree for creating the pid file? Most daemons rely on the fact that the path at which they save their pid exists and its (usually) the administrators responsibility to make sure it exists.

Reply:
I think it's harmless. If the administrator can guarantee the directory, this won't hurt anything. and some of the other daemons do the same thing, such as - docker.

Reply:
:+1: I am generally cautious when applications make too many decisions for me. This helps when faced with unexpected problems. 

Most of the daemons i've seen have this folder creation at their init script (ssh, nginx, apache). 

PS: Sorry if i sound too cautious but i've seen far too many applications fail in insanely unexpected ways :smile:

Reply:
Sometimes, we like the code can do as much as it can to make the great availability, but sometimes not.   In this case, something we probably forgot creates the directory, so we can help to create it, this behavior is the same just like the logs and the data files, but sometimes it might not be, because we configured the incorrect directory. 

And I'm a lazy person, I like the code that can do more works.  ;-)

BTW, Your feedback is very important!  Please keep giving such feedback to me. it's valuable.